libimage/manifest_list.go

```diff
@@ -101,8 +109,161 @@ func (r *Runtime) lookupManifestList(name string) (*Image, manifests.List, error
 	return image, list, nil
 }
 
+// ConvertToManifestList converts the image into a manifest list if it is not
+// already also a list.  An error is returned if the conversion fails.
+func (i *Image) ConvertToManifestList(ctx context.Context) (*ManifestList, error) {
+	// If we don't need to do anything, don't do anything.
+	if list, err := i.ToManifestList(); err == nil || !errors.Is(err, ErrNotAManifestList) {
+		return list, err
+	}
+
+	// Determine which type we prefer for the new manifest list or image index.
+	_, imageManifestType, err := i.Manifest(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("reading the image's manifest: %w", err)
+	}
+	var preferredListType string
+	switch imageManifestType {
+	case manifest.DockerV2Schema2MediaType,
+		manifest.DockerV2Schema1SignedMediaType,
+		manifest.DockerV2Schema1MediaType,
+		manifest.DockerV2ListMediaType:
+		preferredListType = manifest.DockerV2Schema2MediaType
+	case imgspecv1.MediaTypeImageManifest, imgspecv1.MediaTypeImageIndex:
+		preferredListType = imgspecv1.MediaTypeImageIndex
+	default:
+		preferredListType = ""
+	}
+
+	// Create a list and add the image's manifest to it.  Use OCI format
+	// for now.  If we need to convert it to Docker format, we'll do that
+	// while copying it.
+	list := manifests.Create()
+	if _, err := list.Add(ctx, &i.runtime.systemContext, i.storageReference, false); err != nil {
+		return nil, fmt.Errorf("generating new image index: %w", err)
+	}
+	listBytes, err := list.Serialize(imgspecv1.MediaTypeImageIndex)
+	if err != nil {
+		return nil, fmt.Errorf("serializing image index: %w", err)
+	}
+	listDigest, err := manifest.Digest(listBytes)
+	if err != nil {
+		return nil, fmt.Errorf("digesting image index: %w", err)
+	}
+
+	// Build an OCI layout containing the image index as the only item.
```

Comment:
The intent is to force the list/index to be converted to the list format that corresponds to the image type during the `copy.Image()` call further down.  The image library's `storageImageDestination.Commit()` implementation has a special case that writes both the list/index and the image manifest when it notices that the source has multiple instances and it's writing the image's manifest, and we're trying to take advantage of that to add the list/index to the image's set of data files.

Reply:
ACK, I think this does work right now. It’s just that, historically, when we did see c/image failures on mixed OCI/v2s2 data structures (and I forgot the exact details), the response usually was “fix your images, we are not going to add support”.

Assuming c/common will have tests that notice if things break, I’m fine with the mixed-format data here.

---

… uh, actually, wouldn’t it be a bit simpler to use `dir:` instead of `oci/layout` for the top-level source? That would
- not force the top-level item to use the OCI format; we could directly store the output of `list.Serialize`, and expect `c/image.Copy` to record exactly those bytes
- actually be simpler to implement, there are fewer data structures and subdirectories.

But then again, the `oci/layout` producer already exists, so I’m not asking for this to be changed.

---

(AFAICS `sstorageImageDestination.Commit`, ultimately, just adds the multi-platform manifest as a BigData item, which is not all that special — OTOH triggering the `Commit` is valuable in that it triggers _exactly_ the same path as an ordinary pull, i.e. if `Commit` ever changed, this code would continue to work. On balance, I do like triggering the `copy.Image` here, despite all the effort.)

Reply:
I vaguely recall the "dir:" format disclaiming any guarantees about its details not changing going forward, which meant that a change in the image library's reading/writing logic would in turn require updates to logic that created them via other methods as this does, and that possibility being a reason to discourage using it for cases like this.  But it's been a while since I think "dir:" actually last changed, so maybe that's less of a concern now?

Reply:
Yes, the `dir:` format is not formally documented. Pragmatically, we support its use in `skopeo sync`, and that alone implies a fair level of stability over some term of days/weeks(/months?) at least. But then again https://github.com/containers/image/issues/1876 remains outstanding, limiting any possible commitment to _long-term_ stability as of _today_.

An alternative might be to use the `dir` `ImageDestination` to write the data, but writing a top-level manifest without any components is, per the terms of the API, probably (?) undefined (although it would work fine in practice). So that wouldn’t get us to any more _documented_ stability per the state as of _today_.

---

One option is to document things. Probably changing the documentation of `ImageDestination` to explicitly allow creating sparse multi-platform images, if the destination supports them, would be easiest; c/image already has a long-standing RFE to support creating sparse images while mirroring.

---

(Just to be explicit, I’m personally more worried about this code introducing images which are _both_ single-platform “data carrying” images and “multi-platform manifest” images, with a single storage ID; I vaguely seem to remember that the code assumes one or the other. Using mixed-format images for the ~sparse metadata-only copy is a concern, but much less of one, and it can be sufficiently resolved by a small number of tests.)

Reply:
> that the code assumes one or the other

I mean libimage here.